internal/pkg/db/memory/factory.go

```diff
@@ -14,10 +14,49 @@
 
 package memory
 
-type Client struct{}
+import (
+	"sync"
 
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+)
+
+// UnimplementedMethodPanicMessage is the common panic message used for unimplemented methods of memory.Client.
+const UnimplementedMethodPanicMessage = "This method has not been implemented."
+
+// addressableMemoryStore encapsulates the data needed to store addressables.
+type addressableMemoryStore struct {
```

Comment:
Eliminate.

Add global mutex to guard access across entire db/memory implementation.  We shouldn't need a reentrant implementation.

Add addressableMap to Client struct.

Work out how to reuse NewClient in CloseSession; you'll need to capture your mutex before calling NewClient and ensure you unlock the older mutex instance.

Reply:
I originally avoided that approach because I didn't like the idea of the whole client being blocked and possibly preventing unrelated operations from processing. I am fine if we go down that path, I just wanted to bring up the performance impact. Another way would be to create another layer which abstracts locking and ensures access to data which is interconnected to be guarded. That would be more difficult, but much faster.

Reply:
I'm aware of the performance trade-off.  This implementation is specific to acceptance tests only -- for that I'll trade performance for simplicity.

Reply:
Ahhh! I thought this would possibly be used in production, under some odd circumstances. I misunderstood. Thank you for the clarification.